### PR TITLE
DC-195: Cloud Plugin: Deal with RPC Exceptions

### PR DESCRIPTION
**Pull request by patrick-schweizer**
This PR distinguishes between recoverable and non recoverable Exceptions.
see also: https://www.grpc.io/docs/guides/error/

Jira: https://issues.opennms.org/browse/DC-195

### CODE REVIEW

File: plugin/src/main/java/org/opennms/plugins/cloud/tsaas/GrpcExceptionHandler.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+/**
+ * Tries to distinguish between
+ * - recoverable Exceptions: will be propgated to OpenNMS and
+ * - non recoverable Exceptions: will be logged dropped
+ * see also: https://www.grpc.io/docs/guides/error/
+ * */
+public class GrpcExceptionHandler {
+
+  private static final Logger LOG = LoggerFactory.getLogger(GrpcExceptionHandler.class);
+  private static final Set<Code> RECOVERABLE_EXCEPTIONS = new HashSet<>(Arrays.asList(DEADLINE_EXCEEDED, UNAVAILABLE));
```
Comment (patrick-schweizer) [inline plugin/src/main/java/org/opennms/plugins/cloud/tsaas/GrpcExceptionHandler.java:58]: @jbeuree do you think these are the right recoverable exceptions? Sere also: https://www.grpc.io/docs/guides/error/

Comment (jbeuree) [reply plugin/src/main/java/org/opennms/plugins/cloud/tsaas/GrpcExceptionHandler.java:58]: Might make sense to also include the UNAUTHENTICATED error as well - don't drop any outgoing data while we wait for the credentials to be corrected. This should be the error that occurs if we revoke the cert and force them to get a new one.

**Review by jbeuree — CHANGES_REQUESTED** [review]
Just another exception to include, good otherwise.

**Review by jbeuree — DISMISSED** [review]

**Review by freddy61025 — APPROVED** [review]

File: plugin/src/main/java/org/opennms/plugins/cloud/tsaas/GrpcExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+ * - non recoverable Exceptions: will be logged dropped
+ * see also: https://www.grpc.io/docs/guides/error/
+ * */
+public class GrpcExceptionHandler {
+
+  private static final Logger LOG = LoggerFactory.getLogger(GrpcExceptionHandler.class);
+  private static final Set<Code> RECOVERABLE_EXCEPTIONS = new HashSet<>(Arrays.asList(
+          DEADLINE_EXCEEDED,
+          UNAVAILABLE,
+          UNAUTHENTICATED));
```
[the remark this answers is not in the document]
Comment (patrick-schweizer) [reply plugin/src/main/java/org/opennms/plugins/cloud/tsaas/GrpcExceptionHandler.java]: done.

**Pull request merged**